### DISCUSSION TITLE
Jekyll’s `where_exp` workaround

### DISCUSSION THREAD

**Discussion by xplosionmind** [11ty/eleventy#2228] [Q&A]
I am getting crazy converting [my super complex Jekyll website](https://github.com/xplosionmind/tommi.space) to Eleventy. The problem is that I cannot find a way to convert some [Jekyll-only filters](https://jekyllrb.com/docs/liquid/filters/) that I am a heavy user of.

Could you suggest me a way to create a custom filter named `where_exp` in my `.eleventy.js`?  
I would be extremely grateful!

Notes:
- [`where` exists in LiquidJS](https://liquidjs.com/filters/where.html)
- [here](https://github.com/11ty/eleventy/issues/68#issuecomment-383403968) @paulrobertlloyd pointed out the most important Jekyll-only filters that should be supported in Eleventy, but I could not find any solution or alternative.

Thanks again,
Tommi

**Comment by paulrobertlloyd** [comment]
You can register filters in your Eleventy config file, like so:

```js
module.exports = function(eleventyConfig) {
  eleventyConfig.addFilter("where_exp", function(value) {
    // where_exp function
    return value;
  });
};
```

See: https://www.11ty.dev/docs/filters/

**Reply by xplosionmind** [reply]
Exactly, I understand this. What I am asking is the function to be written to replace it… I lack the technical knowledge to write it myself.

**Reply by pdehaan** [reply]
@xplosionmind That could be tricky to implement `where_exp` in JavaScript, since I can't think of an obvious solution that doesn't use the terrible/risky [`eval()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/eval) function.

From the Jekyll docs:
> ### [Where Expression](https://jekyllrb.com/docs/liquid/filters/#where-expression)
> Select all the objects in an array where the expression is true.
> ```liquid
> {{ site.members | where_exp: "item", "item.graduation_year == 2014" }}
> {{ site.members | where_exp: "item", "item.graduation_year < 2014" }}
> {{ site.members | where_exp: "item", "item.projects contains 'foo'" }}
> ```

And even then, `item.projects contains 'foo'` (or even the `nil` keyword) isn't valid JavaScript, so you'd need to do some preprocessing on the data.

And a quick search of npm/Google didn't return any immediate results for an existing Node `where_exp` implementation.
https://24ways.org/2018/turn-jekyll-up-to-eleventy/ had a good overview of converting a couple of Jekyll filters to JavaScript syntax which might be useful.

Probably not the answer you want to hear, but personally I'd create a handful of simple[r] custom filters to do exactly what you need versus trying to write a generic `where_exp` replacement which handles every single case.

A quick [but usually unreliable GitHub] search of your Jekyll blog returned these usages of `where_exp` Jekyll filter in your blog:

```liquid
{% assign apps = site.data.apps | where_exp: 'app', 'app.uninstalled == nil and remember == nil' | sort: 'name' %}
{% assign macosapps = apps | where_exp: 'app', 'app.os contains "MacOS"' | sort: 'name' %}
{% assign alternate = site.pages | concat: site.posts | concat: site.poetry | concat: site.notes | concat: site.filinge | concat: site.sconnesso | where_exp: "entry", "entry.lang != page.lang" | find: "ref", page.ref %}
{% assign accounts = site.data.accounts | where_exp: 'account', 'account.deleted == nil and account.url != nil' %}
{% assign notes = site.notes | where_exp: "note", "note.tags contains 'geek/idea'" | sort: 'updated' | reverse %}
{% assign all = site.posts | concat: site.poetry | concat: site.notes | concat: site.filinge | concat: site.sconnesso | where_exp: 'note', 'note.title != page.title' | sort: 'updated' | reverse %}
{% assign alternate = site.pages | concat: site.posts | concat: site.poetry | concat: site.notes | concat: site.filinge | concat: site.sconnesso | where_exp: "entry", "entry.lang != page.lang" | find: "ref", page.ref %}
```

So instead of this:
```liquid
# Jekyll
{% assign macosapps = apps | where_exp: 'app', 'app.os contains "MacOS"' | sort: 'name' %}
```
maybe write a custom filter like this:

```liquid
# Eleventy ft. LiquidJS
{% assign macosapps = apps | filter_by_os: 'MacOS' | sort: 'name' %}
```

Where your custom filter takes an array of apps and filters results that include the specified OS.
